### PR TITLE
Switch to python 3.11

### PR DESCRIPTION
**Pull request by rajeee**
## Pull Request Description

Black stopped supporting python 3.8: https://github.com/psf/black/pull/4452. We can do the same as well, and make BSB use python 3.11 with support for 3.12. This allows us to use features introduced in later python version such as the [union operator on dict](https://peps.python.org/pep-0584/) and [fine-grained error location](https://docs.python.org/3/whatsnew/3.11.html#whatsnew311-pep657). 
## Checklist

Not all may apply

- [ ] Code changes (must work)
- [ ] Tests exercising your feature/bug fix (check coverage report on Checks -> BuildStockBatch Tests -> Artifacts)
- [ ] Coverage has increased or at least not decreased. Update `minimum_coverage` in `.github/workflows/coverage.yml` as necessary.
- [ ] All other unit and integration tests passing
- [ ] Update validation for project config yaml file changes
- [ ] Update existing documentation
- [ ] Run a small batch run on Kestrel/Eagle to make sure it all works if you made changes that will affect Kestrel/Eagle
- [ ] Add to the changelog_dev.rst file and propose migration text in the pull request

### CODE REVIEW

**Review by nmerket — COMMENTED** [review]
Alternatively we could drop support for python 3.8 altogether and add support for 3.12. I'm able to run on 3.12, but it gives some warnings.

**Comment by rajeee** [comment]
> Alternatively we could drop support for python 3.8 altogether and add support for 3.12. I'm able to run on 3.12, but it gives some warnings.

I considered doing that but remembered your reservation about dropping 3.8 because some part of BSB AWS needing to run inside the openstudio container.

**Review by nmerket — COMMENTED** [review]
Any reason why we're dropping support for 3.9 and 3.10? It's officially supported until October 2025 and 2026, respectively.

Should also have a changelog entry.

**Review by nmerket — APPROVED** [review]

**Review by nmerket — APPROVED** [review]

File: Dockerfile [hunk 1]
```diff
@@ -1,14 +1,17 @@
...
 FROM --platform=linux/amd64 nrel/openstudio:$OS_VER as buildstockbatch
 ARG CLOUD_PLATFORM=aws
-
-RUN sudo apt update && sudo apt install -y python3-pip
-RUN sudo -H pip install --upgrade pip
+ENV DEBIAN_FRONTEND=noninteractive
 COPY . /buildstock-batch/
-RUN python3 -m pip install "/buildstock-batch[${CLOUD_PLATFORM}]"
+
+RUN echo "hello"
```
Comment (nmerket) [inline Dockerfile]:
```suggestion
```

File: Dockerfile [hunk 2]
```diff
@@ -1,14 +1,17 @@
...
+RUN curl -k -LsSf https://astral.sh/uv/install.sh | sh
+ENV PATH="/root/.cargo/bin:$PATH"
+RUN uv python install 3.11
+RUN uv venv --python 3.11 && uv pip install "/buildstock-batch[${CLOUD_PLATFORM}]"
 
 # Base plus custom gems
-FROM buildstockbatch as buildstockbatch-custom-gems
+FROM --platform=linux/amd64 buildstockbatch as buildstockbatch-custom-gems
```
Comment (nmerket) [inline Dockerfile:15]: Ah, was it trying to rebuild that base image on aarch64? Ugh, so annoying.

File: docs/changelog/changelog_dev.rst [hunk 3]
```diff
@@ -117,3 +117,9 @@ Development Changelog
         Updates the bind mount for apptainer to include the whole resources
         directory instead of just the hpxml-measures directory. Makes it work
         with newer versions of ResStock.
+
+    .. change::
+        :tags: general
+        :pullreq: 472
+
+        Remove support for Python 3.8, 3.9, and 3.10.
```
Comment (nmerket) [inline docs/changelog/changelog_dev.rst]:
```suggestion
        Remove support for Python 3.8, 3.9, and 3.10. Add support for Python 3.12.
```

File: setup.py [hunk 4]
```diff
@@ -39,7 +39,7 @@
     long_description_content_type="text/markdown",
     url=metadata["__url__"],
     packages=setuptools.find_packages(),
-    python_requires=">=3.8",
+    python_requires=">=3.11",
```
Comment (nmerket) [inline setup.py]: Does it work with Python 3.13 yet? Should we make that explicit here?

Comment (rajeee) [inline setup.py]:
```suggestion
    python_requires=">=3.11, <3.13",
```

Comment (rajeee) [reply setup.py]: It doesn't. I added that check.

File: Dockerfile (hunk 2 above)
Comment (rajeee) [reply Dockerfile:15]: Yeah, I was getting warnings in my machine.

File: Dockerfile [hunk 5]
```diff
@@ -1,14 +1,16 @@
...
 FROM --platform=linux/amd64 nrel/openstudio:$OS_VER as buildstockbatch
 ARG CLOUD_PLATFORM=aws
-
-RUN sudo apt update && sudo apt install -y python3-pip
-RUN sudo -H pip install --upgrade pip
+ENV DEBIAN_FRONTEND=noninteractive
 COPY . /buildstock-batch/
-RUN python3 -m pip install "/buildstock-batch[${CLOUD_PLATFORM}]"
+
+RUN curl -k -LsSf https://astral.sh/uv/install.sh | sh
```
Comment (rajeee) [inline Dockerfile]:
```suggestion
RUN curl -LsSf https://astral.sh/uv/install.sh | sh
```
Avoid insecure curl.

**Review by nmerket — APPROVED** [review]

**Pull request merged**